ffi/src/utils.rs

```diff
@@ -26,8 +26,26 @@ pub unsafe fn w_str_len(s: *const u16) -> usize {
     len
 }
 
-pub unsafe fn raw_str_into_bytes(raw_buffer: *const c_char, len: usize) -> Vec<u8> {
-    unsafe { from_raw_parts(raw_buffer as *const u8, len) }.to_vec()
+/// Converts raw credentials string into [Vec] of bytes.
+///
+/// Credentials are often represented as strings. For example, username, domain, password.
+/// It is OK for Windows SSPI to accept `null` or empty credential strings. The `AcquireCredentialsHandle`
+/// function will return successful status code is we pass the `null` username value. Thus, this function
+/// will return an empty [Vec] in such a case. It is done on purpose to follow the Windows SSPI behaviour.
+///
+/// # Safety
+///
+/// * `raw_buffer` must be [valid] for reads for `len` many bytes, and it must be properly aligned.
```

Comment:
question: Is `[valid]` instead of `valid` intentional here? I don’t see any reference for that.

Reply:
Nope 😅 
I ran `cargo doc` and found many other warnings. I fixed all of them (see the latest commit: https://github.com/Devolutions/sspi-rs/pull/304/commits/87fc2ab39933c63eac8868515634e24c59957ab2)